File: src/gacha/pool.ts
import type { Student } from "@prisma/client";

export class GachaPool {
  readonly rate: number;

  readonly students: Student[];

  constructor(rate: number) {
    this.rate = rate;
    this.students = [];
  }

  addStudent(student: Student): void {
    this.students.push(student);
  }

  hasStudent(key: string): boolean {
    return this.students.some((student) => student.id === key);
  }

  pull(): [Student, string] | null {
    const index = Math.floor(Math.random() * this.students.length);
    const student = this.students[index];

    if (!student) {
      return null;
    }

    return [student, student.id];
  }
}

File: src/gacha/pool.rs
use crate::db::models::Student;
use rand::Rng;

#[derive(Debug, Clone)]
pub struct GachaPool {
    pub rate: f64,
    pub students: Vec<Student>,
}

impl GachaPool {
    pub fn new(rate: f64) -> Self {
        Self {
            rate,
            students: Vec::new(),
        }
    }

    pub fn add_student(&mut self, student: Student) {
        self.students.push(student);
    }

    pub fn has_student(&self, key: &str) -> bool {
        self.students.iter().any(|s| s.id == key)
    }

    pub fn pull<R: Rng + ?Sized>(&self, rng: &mut R) -> Option<(Student, String)> {
        if self.students.is_empty() {
            return None;
        }
        let index = rng.gen_range(0..self.students.len());
        let student = self.students[index].clone();
        let id = student.id.clone();
        Some((student, id))
    }
}
